Bottom: 416dfe20fb5b032eaa93cf78cb48fb9feab202ac
Top:    8da307aebb8d386f5b764622a5d85fc9f248fa62
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-01-22 18:42:14 +0800



---

diff --git a/drivers/vhost/vhost.c b/drivers/vhost/vhost.c
index 088afe5..893cad7 100644
--- a/drivers/vhost/vhost.c
+++ b/drivers/vhost/vhost.c
@@ -2414,9 +2414,9 @@ struct vhost_msg_node *vhost_dequeue_msg(struct vhost_dev *dev,
 }
 EXPORT_SYMBOL_GPL(vhost_dequeue_msg);
 
-int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
-				struct vring_used_elem *heads,
-				u16 num)
+static int vhost_read_desc_indices(struct vhost_virtqueue *vq,
+				   struct vring_used_elem *heads,
+				   u16 num)
 {
 	int ret, ret2;
 	u16 last_avail_idx, last_used_idx, total, copied;
@@ -2444,28 +2444,10 @@ int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
 		last_avail_idx = (last_avail_idx + 1) & (vq->num - 1);
 	}
 
-	last_used_idx = vq->last_used_idx & (vq->num - 1);
-	while (total) {
-		copied = min((u16)(vq->num - last_used_idx), total);
-		ret2 = vhost_copy_to_user(vq,
-					  &vq->used->ring[last_used_idx],
-					  &heads[ret - total],
-					  copied * sizeof(*used));
-
-		if (unlikely(ret2)) {
-			vq_err(vq, "Failed to update used ring!\n");
-			return -EFAULT;
-		}
-
-		last_used_idx = 0;
-		total -= copied;
-	}
-
 	/* Only get avail ring entries after they have been exposed by guest. */
 	smp_rmb();
 	return ret;
 }
-EXPORT_SYMBOL(vhost_prefetch_desc_indices);
 
 static int __init vhost_init(void)
 {
diff --git a/drivers/vhost/vhost.h b/drivers/vhost/vhost.h
index 1527f5c..79c6e7a 100644
--- a/drivers/vhost/vhost.h
+++ b/drivers/vhost/vhost.h
@@ -224,9 +224,6 @@ ssize_t vhost_chr_read_iter(struct vhost_dev *dev, struct iov_iter *to,
 ssize_t vhost_chr_write_iter(struct vhost_dev *dev,
 			     struct iov_iter *from);
 int vhost_init_device_iotlb(struct vhost_dev *d, bool enabled);
-int vhost_prefetch_desc_indices(struct vhost_virtqueue *vq,
-				struct vring_used_elem *heads,
-				u16 num);
 
 #define vq_err(vq, fmt, ...) do {                                  \
 		pr_debug(pr_fmt(fmt), ##__VA_ARGS__);       \
